blueprints/addon/index.js

```diff
@@ -77,6 +82,19 @@ module.exports = {
     // 100% of addons don't need ember-cli-app-version, make it opt-in instead
     delete contents.devDependencies['ember-cli-app-version'];
 
+    // add scripts to build type declarations for TypeScript addons
+    if (this.options.typescript) {
+      contents.scripts.prepack = 'tsc --project tsconfig.declarations.json';
+      contents.scripts.postpack = 'rimraf declarations';
+
+      contents.typesVersions = {
+        '*': {
+          'test-support/*': ['declarations/addon-test-support/*'],
```

Comment:
This is doing Ember's conventional mapping of `addon-test-support` (author) → `test-support` (consumer).

I have seen some issues though importing things from index-modules using that `typesVersions` approach, like having a `addon/foo/index.ts` module and trying to import that as `import foo from 'my-addon/foo'`, which did cause type errors, but `import foo from 'my-addon/foo/index'` did not. This seems to be some [bug in TypeScript itself](https://github.com/microsoft/TypeScript/issues/43133), but maybe someone here has a suggestion how to get around this? 

That's where definitely some more testing is needed...

Reply:
> here has a suggestion how to get around this?

```
"*": {
  "test-support": ["declarations/addon-test-support/index.d.ts"]
}
```
should cover that?

Reply:
What about arbitrary modules you might want to export? Like `import foo from 'my-addon/foo/bar/baz'`

However coincidentally I was running into that also in another project, and we got it working _I think_ with something like that:

```
{
        '*': {
          'test-support/*': ['declarations/addon-test-support/*', 'declarations/addon-test-support/*/index.d.ts'],
          '*': ['declarations/addon/*', 'declarations/addon/*/index.d.ts'],
        },
```

- [ ] I need to double check that though...

Reply:
I tested different variations of imports, and this is what seems to be working fine now:

```
  "typesVersions": {
    "*": {
      "test-support": [
        "declarations/addon-test-support/index.d.ts"
      ],
      "test-support/*": [
        "declarations/addon-test-support/*",
        "declarations/addon-test-support/*/index.d.ts"
      ],
      "*": [
        "declarations/addon/*",
        "declarations/addon/*/index.d.ts"
      ]
    }
  }
```